### ISSUE TITLE
Please explain why I need this

### ISSUE THREAD

**Issue by fractalf** [request/request-promise#121]
The description says `The simplified HTTP request client 'request' with Promise support. Powered by 'bluebird'.`

I don't think you can get more simple than this:
````
return new Promise(function (fulfill, reject) {
    request(options, function (error, response, html) {
        if (error !== null) {
            reject(error);
        } else {
            fulfill(html);
        }
    });
});

````

What else does this lib offer?

I'm lazy, I didn't read past the description, and that didn't tell me anything

**Comment by analog-nico** [comment]
Please be so kind and read the README first. If you still have questions I am happy to help.

**Comment by fractalf** [comment]
I did read quite far down and couldn't see anything other than a wrapper for request, so this was a (bad?) hint on maybe put some short info about it early on. The README is very long, one of those you skim until you deside to go for it.

So anyway, just a hint. Close at will :)

**Comment by analog-nico** [comment]
Ok, I see. Thanks for your hint!

I'll keep this issue open until I redid the docs.

My current thoughts as a starting point for creating a better pitch:

- Request-Promise is mostly for convenience and to reduce boilerplate code
  - Since it exposes Request's full API the boilerplate code reduction can be big if you use different kinds of requests. E.g. some with different default parameters (using `.defaults(...)`) or with extra setup like cookie handling (using `.jar()` and `.setCookie(...)`).
  - The default configuration `simple = true, resolveWithFullResponse = false` provides a very fitting way to work with Promises. Less if's etc. if the request is part of a larger Promise chain.
  - The `transform` option plus using `.defaults(...)` adds extra convenience if your code uses multiple common ways to process responses.
- There are a few features coming up that add extra value:
  - Cancellation support https://github.com/request/request-promise/issues/113
  - Allow resolving promise with stream https://github.com/request/request-promise/issues/90

**Comment by fractalf** [comment]
Hi again,
thanks for the draft for a better pitch. It's better, but to be honest what I was looking for when I started reading the README was a brief paragraf/pointed list of things I get using this lib that my little code snip above don't give me. I consider myself a somewhat intermediate developer. A lot of the examples (at least on the top) just look like a wrapper to the old request lib, plus that little added Promise.
Thanks for the effort

**Comment by analog-nico** [comment]
Quick question: Do you argue from the standpoint that you think better code snippets can be found? (Suggestions welcome.) Or do you expect request-promise to be more than just a wrapper? (Sorry, it is just a wrapper.)

**Comment by fractalf** [comment]
No, I'm not trying to critizise the code itself and I do understand that it's just a wrapper.
Here's exactly what going on. A couple of months ago I needed request to return a promise, so I basically just wrapped it into one using Bluebird like I showed in the first post. It solved my problems. Just now recently I picked up this project again and thought to look if request had gotten promise support. It hasn't and I stumbled upon this repo. I tried to read the source code, but couldn't understand what was going on, and I couldn't understand what was extra stuff I would get using this instead of what I had made. I got confused when I read the docs, cause some of it looked like just request usage, I wanted the "diff".

**Comment by analog-nico** [comment]
Ok, I understand now. I'll try to improve on that.

**Comment by olsonpm** [comment]
@fractalf - I had the same confusion until I tried wrapping request via bluebird's promisifyAll (which I try to do for every async library by default).  I can't remember whether it was request's optional callback parameter or default streaming support, but I was unable to get promisifyAll working which is when I searched for a pre-built promisified version of request.

I understand this doesn't help solve the problem for request-promise in particular, but maybe gives future readers a different heuristic for deciding when a promisified wrapper to a library is necessary.

**Comment by fractalf** [comment]
@olsonpm  - I'm not sure we share the same confusion here? :) I know the original request needs to be wrapped to make it work with promises. I rarely use promisifyAll as I find it easier to understand what's going on by making my own wrapper. So "my" confusion is why I need more than what I posted in the first post. The basics of Promises/A+ is quite straight forward

**Comment by olsonpm** [comment]
To be clear, I assumed our shared confusion is why we need a wrapper library if the transition is extremely simple.

I still think the above applies, we just define "simple" differently.  I define it by being able to use promisify and you define it by your snippet.  My question is then, why not always start with the snippets since depending on a wrapper lib is usually not the greatest idea to begin with?  Use the wrapper when managing your translations become a hassle.

**Comment by MarkHerhold** [comment]
I do not agree with @fractalf 's issue around documentation clarity.

I always thought that the README was perfectly clear on what what the purpose of the module was and the features it added. I think anyone that has used node's built in `http` module or even the plain `request` and also needed promises should be able to see the value in this module immediately.

I think that the current readme is brilliantly simple in that it tells developers exactly what this module is doing for them right away and even highlights the fundamental differences from the `request` module.

It's hard to get more simple than this. This statement clearly describes most everyone's first foray into trying to promisify request and clearly states some key functionality that devs would find immediately useful.

> Request and Bluebird are pretty awesome, but I found myself using the same design pattern. Request-Promise adds a Bluebird-powered .then(...) method to Request call objects. By default, http response codes other than 2xx will cause the promise to be rejected. This can be overwritten by setting options.simple = false.

**Comment by olsonpm** [comment]
@MarkHerhold - Because bluebird advises 'promisifyAll' and 'promisify', this library should be unnecessary.  It's necessary however because the request library doesn't or can't structure its api to support promisify.  This means a workaround is necessary.  Instead of everyone working around via their own promisified version of 'request', this library allows all users to share a single workaround implementation.

The docs don't mention this.

**Comment by analog-nico** [comment]
Hi @MarkHerhold and @olsonpm you both got good points here. I think it depends on where you come from in regard to what info you are looking from:

- First-time `request` user or `request` + callback user -> I agree with @MarkHerhold that the README does a good job at that
- Request user – like @olsonpm – with your own promise wrapping / promisifyAll -> A section could be useful to explain the differences compared to `request-promise` – see #96 btw

Although @fractalf and @olsonpm have a somewhat different perspective the challenge obviously is to already cover their perspectives in the pitch for the library at the very beginning.

Actually, we are working on a new documentation website for `request` and `request-promise`: http://request.js.org ([repo](https://github.com/request/request.github.io)) There is not much on there yet but at least we have the [Getting Started section](http://request.js.org/docs/getting-started.html). It would be great if you could give some initial feedback and how to structure the sections that will come afterwards. Thanks!

**Comment by olsonpm** [comment]
Thanks for the response.  I'm willing to give initial feedback on the [request-promise docs](http://request.js.org/docs/request-promise/) once it's there, and I think as long as you

1 - introduce the complexity of converting the request api to its promise counterpart (to cover fractalf's question)
2 - give a brief explanation on why promisifyAll doesn't work with request (to cover my initial question)

then all will be well

**Comment by analog-nico** [comment]
Alright, thanks @olsonpm !
